src/Featurizers/BackwardFillImputerFeaturizer.h

```diff
@@ -0,0 +1,224 @@
+    // |
+    // ----------------------------------------------------------------------
+    BackwardFillImputerTransformer(void);
+    BackwardFillImputerTransformer(T defaultValue);
+    BackwardFillImputerTransformer(Archive &ar);
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(BackwardFillImputerTransformer);
+
+    bool has_pending(void) const;
+    BackwardFillImputerTransformer & reset(void);
+
+    void save(Archive &ar) const override;
+
+    bool operator==(BackwardFillImputerTransformer const &other) const;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Types
+    // |
+    // ----------------------------------------------------------------------
+    using NullableType                      = nonstd::optional<T>;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    // We are using nonstd::optional here, as we always want the type to be optional
+    // and not result is default values which my not be optional (for example, double
+    // is technically optional (as it supports NaN), but default-constructs as something
+    // that isn't NaN; we avoid this problem by always using optional).
+    NullableType const                      _defaultValue;
+    std::uint32_t                           _numPending;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC runs into problems when the definition and declaration are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        // ----------------------------------------------------------------------
+        using TheseTraits                   = Traits<typename BaseType::InputType>;
+        // ----------------------------------------------------------------------
+
+        if(TheseTraits::IsNull(input)) {
+            ++_numPending;
+            return;
+        }
+
+        T const &                           value(TheseTraits::GetNullableValue(input));
+        std::uint32_t                       toWrite(_numPending + 1);
+
+        while(toWrite--)
+            callback(value);
+
+        _numPending = 0;
+    }
+
+    void flush_impl(typename BaseType::CallbackFunction const &callback) override {
+        // ----------------------------------------------------------------------
+        using NullableTypeTraits            = Traits<NullableType>;
+        // ----------------------------------------------------------------------
+
+        if(_numPending != 0) {
+            if(NullableTypeTraits::IsNull(_defaultValue))
+                throw std::runtime_error("Pending backward fill items remain");

```

Comment:
Is there any way to suppress this error and let it continue? This is one of the use cases AutoML told us is the reasons that they will use multiple imputation strategies, so that if any of them still have nulls that another imputation strategy will catch it.

Reply:
Similar comment as the other - I've added a TODO but we need to think about the scenario more before implementing the solution.

---
In reply to: [383484002](https://github.com/microsoft/FeaturizersLibrary/pull/144#discussion_r383484002) [](ancestors = 383484002)